src/OpenTelemetry/Metrics/MetricReaderExt.cs

```diff
@@ -264,25 +292,20 @@ private Batch<Metric> GetMetricsBatch()
             int metricCountCurrentBatch = 0;
             for (int i = 0; i < target; i++)
             {
-                var metric = this.metrics![i];
-                int metricPointSize = 0;
+                ref var metric = ref this.metrics![i];
                 if (metric != null)
                 {
-                    if (metric.InstrumentDisposed)
-                    {
-                        metricPointSize = metric.Snapshot();
-                        this.instrumentIdentityToMetric.TryRemove(metric.InstrumentIdentity, out var _);
-                        this.metrics[i] = null;
-                    }
-                    else
-                    {
-                        metricPointSize = metric.Snapshot();
-                    }
+                    int metricPointSize = metric.Snapshot();
 
                     if (metricPointSize > 0)
                     {
                         this.metricsCurrentBatch![metricCountCurrentBatch++] = metric;
                     }
+
+                    if (!metric.Active)
```

Comment:
We should move this up and do this right after `int metricPointSize = metric.Snapshot();` like the original code. That's just to reduce the likelihood of someone running into the race condition where `instrumentIdentityToMetric` returns a Metric that is about to be marked as `null`.

Reply:
I made it `ref metric` so we could clear the array storage without re-indexing. See how on line 302 it sets the metric into the batch array? If I move the clear above line 302 that ref may be null so we would need to take a copy of it. Switching the order kept it fast. I don't think we need to worry about the race any longer though because https://github.com/open-telemetry/opentelemetry-dotnet/pull/4958/files#diff-8650a8083e99ad454eac9d5bef1a6061cc0c1ab31085a1d25e5c4f57ed16f74fR260-R261 takes care of it.